keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -369,12 +376,27 @@ of the following scenarios:
 2. A new Node results in a Service that is able to achieve an endpoint
    distribution below 20% for the first time.
 
+### SameZone Heuristic
+The simplest of all heuristics, this will simply copy the zone of an endpoint to
+the hint of the same endpoint. If there are not endpoints available within a zone,
+that will mean no endpoints will have hints for that zone. As described above,
+when Kube-Proxy cannot find any endpoints with hints for its zone, it will route
+traffic ~evenly across all endpoints, regardless of hints.
+
+### Additional Heuristics
+To enable additional heuristics to be added in the future, we will:
+
+1. Remove the requirement in kube-proxy that the hints annotation must be set to
+   a known value on the associated Service before the values of EndpointSlice
+   hints will be considered.
+2. Ensure the EndpointSlice controller TopologyCache provides an interface that
+   simplifies adding additional heuristics in the future.
+
 ### Future Expansion
 
 In the future we may expand this functionality if needed. This could include:
 
-- A new `RequireZone` algorithm that would keep endpoints in EndpointSlices for
-  the same zone they are in.
+- As described above, additional heuristics may be added in the future.
```

Comment:
Before I start - a SameZone heuristic absolutely fixes my main issues with the current Hints implementation in that it gives me a consistent way to enforce intra-zonal service routing, and if we went ahead with this as proposed it would make me a happy camper.

That said - do you think we could **also** add a "nearest" heuristic? Either in this PR or future (ignore my naming though). The idea would be to replicate the old topology keys behaviour of:
```
  topologyKeys:
    - "kubernetes.io/hostname"
    - "topology.kubernetes.io/zone"
    - "topology.kubernetes.io/region"
    - "*"
```

We use this probably 99% of the time as it allows us to:
- Run things like CoreDNS as a daemonset and ensure that 99% of the time DNS traffic is kept local, but if the local pod is down still serve DNS.
- We run bare metal and expose our ingress controllers (and other apps) over BGP using MetalLB, with the ingress controller scheduled to run on the the same nodes as MetalLB. The above setup means we can avoid inter-nodal traffic (`ExternalIP -> ClusterIP -> PodIP` on a different node). We then rely on ECMP externally to ensure we're load balancing across the various nodes.

Between `nearest` and `SameZone` I feel like it fits all my use cases..

Reply:
"region" isn't super meaningful because (unless you define region differently) most clusters do not span regions.  If that assumption isn't true, I'd like to discuss it :)

What you are asking for is a "SameNodeSameZone" sort of thing, which is the logical extension of this.  Except we don't have a `forNode` hint, so the implementation would be different.

Reply:
My clusters do span regions and we do make an important distinction between Zone and Region, namely Zones correspond to physical location and Regions to geographical areas (for a cloud analogy, AZs and Regions). For example we have a cluster that looks like:
```
<Region>
  - <Zone>
     - <Node>
NYC:
  - DC1
     - Node 1
     - Node 2
     - <etc.>
  - DC2
  - DC3
IND:
  - DC1
  - DC2
LON:
  - DC1
```

We can then do things with scheduling such that all applications by default have replicas in:
 - At least 2 regions (we have a mix of "global" and region specific applications, based on storage and/or user proximity)
 - At least 2 zones within the region (if possible)
 - At least 2 nodes within the zone

(the fact that the above can all be done with default scheduling profiles is awesome)

Anyway - I want `SameNodeSameZoneSameRegion`, I hope that explains why.

Reply:
> My clusters do span regions

Yikes!  There are a couple of places where we kind of assume they don't.  The example you cite is not something we design or test for.  there's an assumption that the control-plane nodes are close to each other and that all nodes are "close" to the control plane.  If you had said this was all one facility, but different buildings mapped to  regions, and different racks mapped to zones I might be less scared.

I am a little surprised it works...

Reply:
indeed, assuming one etcd instance per region, the latency between LON-NY per example for etcd has to be a big impact, once you have a big churn of operations or a large number of objects your cluster is going to suffer a lot, not mentioning the high risk of split brain scenarios that setup exhibits.

 @thockin @robscott these power users will demand more and more heuristics, or some heuristics that need match their setups, that is why I think we should make a TopologyHints controller out of tree or some way https://github.com/kubernetes/enhancements/pull/3765#discussion_r1083116250 the people can use to implement their own heuristics.

Reply:
Sorry I wasn't clear, but those assumptions aren't right.

I run ETCD bare-metal completely separately from Kubernetes, it runs across multiple datacenters within 2 geographically close regions where we have dark fiber connectivity, so latency is <10-20ms. Kubernetes control plane components follow a similar setup, though API servers are distributed more widely.

It's the workers that are distributed truly globally, but again we have dedicated WAN connectivity such that even HK<->NYC is at worst 250ms. I've not seen any noticeable issues with the worker<->API latencies.

Reply:
I'm glad to hear it works for you, I am still a little scared :)

I'll take it as a personal mission to stop assuming single-region :)